repose-aggregator/tests/functional-tests/src/integrationTest/groovy/features/filters/apivalidator/ApiValidatorMultiTenantTest.groovy

```diff
@@ -0,0 +1,475 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=_
+ */
+package features.filters.apivalidator
+
+import org.openrepose.framework.test.ReposeValveTest
+import org.rackspace.deproxy.Deproxy
+import org.rackspace.deproxy.MessageChain
+import spock.lang.Unroll
+
+import static javax.servlet.http.HttpServletResponse.SC_FORBIDDEN
+import static javax.servlet.http.HttpServletResponse.SC_OK
+import static org.openrepose.commons.utils.http.OpenStackServiceHeader.ROLES
+import static org.openrepose.commons.utils.http.PowerApiHeader.RELEVANT_ROLES
+import static org.openrepose.commons.utils.string.Base64Helper.base64EncodeUtf8
+
+class ApiValidatorMultiTenantTest extends ReposeValveTest {
+
+    static def mapHeaderValue = base64EncodeUtf8("""{
+      |   "1" : ["a:admin","foo","bar"],
+      |   "2" : ["a:creator", "foo", "a:observer"],
+      |   "3" : ["a:updater", "bar", "biz", "a:creator"],
+      |   "4" : ["a:observer"],
+      |   "5" : ["a:admin", "bar", "biz", "a:creator"],
+      |   "6" : ["biz", "baz"],
+      |   "7" : ["role with spaces", "biz"],
+      |   "8" : ["a:patcher"]
+      |}
+      |""".stripMargin())
+
+    static def observerRequests = [
+        ["GET", "/v1/resource"]
+    ]
+
+    static def creatorRequests = [
+        ["POST", "/v1/resource"],
+        ["POST", "/v1/resource/other"]
+    ]
+    static def updaterRequests = [
+        ["PUT", "/v1/resource"],
+        ["PUT", "/v1/resource/other"]
+    ]
+    static def adminOnlyRequests = [
+        ["DELETE", "/v1/resource"],
+        ["DELETE", "/v1/resource/other"]
+    ]
+
+
+    def setupSpec() {
+        deproxy = new Deproxy()
+        deproxy.addEndpoint(properties.targetPort)
+
+        def params = properties.getDefaultTemplateParams()
+        repose.configurationProvider.applyConfigs("common", params)
+        repose.configurationProvider.applyConfigs("features/filters/apivalidator/multitenant", params)
+        repose.start()
+
+        repose.waitForNon500FromUrl(reposeEndpoint)
+    }
+
+    @Unroll("Should succeed on GET /v1/resource/other when appropriate headers are set (#tenants)")
+    def "Should succeed when appropriate headers are set"() {
+        when:
+        MessageChain mc = deproxy.makeRequest(
+            url: "$reposeEndpoint/v1/resource/other",
+            method: "GET",
+            headers: [
+                "X-Tenant"   : tenants,
+                "X-Other"    : "other",
+                (ROLES)      : "foo,bar",
+                "X-Map-Roles": mapHeaderValue])
```

Comment:
This should probably use the constant like `ROLES`. So `OpenStackServiceHeader.TENANT_ROLES_MAP`.

Reply:
I looked for it, but I didn't find it.
Now it makes sense. It is on your branch that isn't on master yet.

Reply:
This header wasn't added in it's own commit, so I can't just `cherry-pick` it.

Reply:
Nope. You would have to rebase, I guess. Or I could break that constant out into its own PR...

But since tests just passed on the branch, it should be good to go after one more quick glance form Adrian. So maybe just wait for that? Or don't use the constant, but I'm really not a fan of that.

Reply:
Yep. Sounds good. Hopefully it will be in shortly.